### PR TITLE
[swl-6332] bundle task supports multiple outstanding operations

### PR DESCRIPTION
**Pull request by archerhungbsn**
Reviewer: @kenchiang @poolakiran @wilmo119 
Support multiple outstanding operations.

### CODE REVIEW

File: modules/OFConnectionManager2/utest/main.c [hunk 1]
```diff
@@ -1093,6 +1096,43 @@ subbundle_finishes[] = {
     subbundle_finish,
 };
 
+uint32_t invoke_pre_start[NUM_SUBBUNDLES];
+uint32_t invoke_post_finish[NUM_SUBBUNDLES];
```
Comment (kenchiang) [inline modules/OFConnectionManager2/utest/main.c]: even though this is just the utest, suggest renaming "pre_start" to "start3" and "post_finish" to "finish3" throughout the utest.  "pre_start" implies one step before "start", and "post_finish" implies one step after "finish", whereas i think we just want names that don't overlap with the existing names from testing set2

File: modules/OFConnectionManager2/utest/main.c [hunk 2]
```diff
@@ -1093,6 +1096,43 @@ subbundle_finishes[] = {
...
+
+void
+subbundle_pre_start(indigo_cxn_id_t cxn_id, indigo_cxn_subbundle_info_t *subbundle_info)
+{
+    uint32_t subbundle_idx = subbundle_info->subbundle_idx;
+    printf("pre_setart subbundle %u\n", subbundle_idx);
```
Comment (kenchiang) [inline modules/OFConnectionManager2/utest/main.c]: "start", not "setart"

File: modules/OFConnectionManager2/utest/main.c [hunk 3]
```diff
@@ -1494,7 +1594,13 @@ test_async_op(char *controller_suffix,
     INDIGO_ASSERT(obj->object_id == OF_BARRIER_REPLY,
                   "did not receive OF_BARRIER_REPLY, got %d", obj->object_id);
 
-    /* bundle add test: open, add echoes, commit */
+    /* Test case: bundle async test
+     * steps:
+     *     open, add echo, add async pending, add echo, commit
+     * expect:
+     *     only first echo reply, then wait for unblock
+     *     second echo rply after unblock
```
Comment (kenchiang) [inline modules/OFConnectionManager2/utest/main.c:1612]: "rply" -> "reply"

File: modules/OFStateManager/module/src/gentable_handlers.c [hunk 4]
```diff
@@ -31,8 +31,8 @@
  * Async Operations:
  * The add4()/modify4()/del4() will support async operations.
  * Driver:
- *    The return value, INDIGO_ERROR_PENDING, indicates that driver
- *    is performing an async operation.
+ *    The return value, INDIGO_ERROR_PENDING or INDIGO_ERROR_CONTINUE,
```
Comment (kenchiang) [inline modules/OFStateManager/module/src/gentable_handlers.c:34]: "A return value of INDIGO_ERROR_PENDING or INDIGO_ERROR_CONTINUE indicates..."

File: modules/OFStateManager/module/src/gentable_handlers.c [hunk 5]
```diff
@@ -51,7 +51,7 @@
  *    operation.
  *
  *    Only add4()/modify()/del4() will expect INDIGO_ERROR_PENDING
- *    return.
+ *    or INDIGO_ERROR_CONTINUE return.
```
Comment (kenchiang) [inline modules/OFStateManager/module/src/gentable_handlers.c:54]: Only add4/modify4/del4 will return INDIGO_ERROR_PENDING or INDIGO_ERROR_CONTINUE

File: modules/indigo/module/inc/indigo/of_connection_manager.h [hunk 6]
```diff
@@ -814,6 +814,49 @@ indigo_cxn_subbundle_set2(uint32_t num_subbundles,
...
+
+typedef void (*indigo_cxn_subbundle_pre_start_t)(indigo_cxn_id_t cxn_id,
+                                                 indigo_cxn_subbundle_info_t *subbundle_info);
+typedef void (*indigo_cxn_subbundle_post_finish_t)(indigo_cxn_id_t cxn_id,
+                                                   indigo_cxn_subbundle_info_t *subbundle_info);
+
+/**
+ * Extends the subbundle infrastructure from indigo_cxn_subbundle_set2 above.
+ * First five parameters are the same as for indigo_cxn_subbundle_set2.
+ * @param pre_starts Array of prestart functions; the ith function is called
```
Comment (kenchiang) [inline modules/indigo/module/inc/indigo/of_connection_manager.h]: let's clean up all instances of the prestart/postfinish language, which will just serve to confuse

File: modules/indigo/module/inc/indigo/of_connection_manager.h [hunk 7]
```diff
@@ -814,6 +814,49 @@ indigo_cxn_subbundle_set2(uint32_t num_subbundles,
...
+ * before the ith subbundle is started.
+ * @param post_finishes Array of postfinish functions; the ith function is called
+ * after the ith subbundle is finished.
+ *
+ * The intent is to provide subbundle information to the the flowtables or
+ * gentables before or after processing the subbundle.
+ * Adding these two functions also try to backward compatible the existing
+ * codes instead of modifying the start() and finish() functions.
```
Comment (kenchiang) [inline modules/indigo/module/inc/indigo/of_connection_manager.h:846]: let's just delete lines 845-846, and replace with "call only one of indigo_cxn_subbundle_set, indigo_cxn_subbundle_set2, or indigo_cxn_subbundle_set3 functions"

**Review by kenchiang — COMMENTED** [review]
This is looking good!  my suggestions are to (1) clean up the leftover prestart/postfinish names and comments and (2) add some more comments describing the use of the PENDING and CONTINUE return codes.

In particular, we need to call out in the header file comments that PENDING will pause bundle processing, and that both PENDING and CONTINUE will increment the pending count.  So the driver needs to call the resume API once for each PENDING or CONTINUE return code that the driver sends.

I may be misunderstood something, so please feel to correct me.  Thanks.

File: modules/indigo/module/inc/indigo/of_connection_manager.h (hunk 6 above)
Comment (archerhungbsn) [reply modules/indigo/module/inc/indigo/of_connection_manager.h]: not sure why you still see the stale codes. They have been removed

**Comment by kenchiang** [comment]
i reviewed last two commits after we spoke yesterday (370dbc and b2d9c0f), and they look good.

**Pull request merged**